### PR TITLE
[Model] Apertus 1.5 multimodality

### PR DESCRIPTION
**Pull request by blancsw**
## General Info

This PR is the Apertus 1.5 implementation with image and audio multimodality. The model weights will be available very soon.

Special thanks to @tomas-polach for the bug fix contribution!
And @Anunay-Yadav for his startpoint code at https://github.com/swiss-ai/vllm/tree/apertus_integration

## Purpose

This PR introduces native vLLM asynchronous execution support for the multimodal **Apertus 1.5** model, integrating **Emu3.5** for vision tokenization and **WavTokenizer40** for audio tokenization.

The implementation strictly adheres to a high-performance CPU/GPU separation architecture contract to optimize execution and avoid pipeline stalls:
- **Processor (CPU-bound, zero NN inference):** Implements deterministic tensor creation, dynamic batch padding, and dummy token injection. Variable-resolution images and variable-length audio sequences are mathematically padded to the maximum dimensions in the batch (`apertus_utils.py`, `ApertusMultiModalProcessor` in `apertus_mm.py`).
- **Worker (GPU-bound, native execution):** Heavy neural network inference executes natively on the GPU (`ApertusForConditionalGeneration` in `apertus_mm.py`). It embeds pixel values using the Emu3.5 vision tower, embeds audio using WavTokenizer, and performs O(1) boolean mask ID substitution to inject visual/audio embeddings directly into the language model.
- **Self-contained Tokenizers:** Introduces clean, self-contained, and statically-instantiated versions of `Emu3.5` (`apertus_emu35.py`) and `WavTokenizer40` (`apertus_wavetokenizer.py`). This eliminates dynamic imports and external path injections, ensuring stability and simplified deployment.

## Test Plan

```bash
vllm serve apertus-ai/Apertus-v1.5-70B-Prerelease-2607 \
  --enable-chunked-prefill \
  --async-scheduling \
  --enable-prefix-caching \
  --served-model-name \
    apertus-ai/Apertus-v1.5-70B-Prerelease-2607 \
  --tensor-parallel-size 2 \
  --quantization fp8 \
  --kv-cache-dtype fp8 \
  --max-model-len 100000 \
  --trust-remote-code \
  --max-num-seqs 128 \
  --max-num-batched-tokens 65536 \
  --block-size 16 \
  --enable-auto-tool-choice \
  --tool-call-parser apertus \
  --hf-overrides '{"architectures": ["ApertusForConditionalGeneration"]}'
```

## Test Result

TODO

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/apertus.py [hunk 1]
```diff
@@ -73,6 +74,8 @@
     maybe_prefix,
 )
 
+logger = init_logger(__name__)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/apertus.py]: Seems unused

File: vllm/model_executor/models/apertus_emu35.py [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+            full_zeros[:, self.used, ...] = soft_one_hot
+            soft_one_hot = full_zeros
+
+        dim = 1
+        ind = soft_one_hot.max(dim, keepdim=True)[1]
+        hard_one_hot = torch.zeros_like(
+            logits, memory_format=torch.legacy_contiguous_format
+        ).scatter_(dim, ind, 1.0)
+
+        if self.training:
```
Comment (DarkLight1337) [inline vllm/model_executor/models/apertus_emu35.py]: Please remove the training specific code

Comment (blancsw) [reply vllm/model_executor/models/apertus_emu35.py]: Thank you for the review! I am working on it now.

File: vllm/model_executor/models/apertus_emu35.py [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+        self.batch_maximization_weight = batch_maximization_weight
+
+        self.l2_normalize = l2_normalize
+        if self.l2_normalize:
+            self.init_embedding()
+
+    def init_embedding(self):
+        embedding = torch.randn(self.n_e, self.e_dim)
+        embedding = embedding / embedding.norm(dim=-1, keepdim=True).clamp(min=1e-8)
+        self.embedding.weight.data = embedding
```
Comment (Isotr0py) [inline vllm/model_executor/models/apertus_emu35.py]:
```suggestion
```
No need for inference.

File: vllm/model_executor/models/apertus_mm.py [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+
+            logger.info(
+                "[Apertus Worker] Loading Audio Tower natively on %s",
+                target_device,
+            )
+            self.audio_tower = self.audio_tokenizer.load_audio_tokenizer(
+                model_path=model_path,
+                device=str(target_device),
+                audio_config=self.config.audio_config,
+            )
```
Comment (Isotr0py) [inline vllm/model_executor/models/apertus_mm.py:450]: This is quite messy, please use `AutoWeightsLoader` and `DefaultModelLoader.Source`
https://github.com/vllm-project/vllm/blob/ce653856186e80b2a927a611cb184aeac64d2862/vllm/model_executor/models/kimi_audio.py#L404-L410
https://github.com/vllm-project/vllm/blob/ce653856186e80b2a927a611cb184aeac64d2862/vllm/model_executor/models/kimi_audio.py#L575-L590

File: vllm/model_executor/models/apertus_mm.py [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+                # Convert WavTokenizer codes to language model vocabulary space
+                llm_audio_ids_i = valid_codes_i.to(torch.long) + self.audio_token_offset
+
+                # Fetch embeddings from text embedding layer for the resolved token IDs
+                audio_embeds_i = super().embed_input_ids(llm_audio_ids_i.to(device))
+                valid_codes_embeds.append(audio_embeds_i)
+
+            return valid_codes_embeds
+
+        return []
```
Comment (Isotr0py) [inline vllm/model_executor/models/apertus_mm.py]: You can refer to `qwen3_vl.py` to implement `embed_multimodal` cleaner:
https://github.com/vllm-project/vllm/blob/ce653856186e80b2a927a611cb184aeac64d2862/vllm/model_executor/models/qwen3_vl.py#L2747-L2775

File: vllm/model_executor/models/apertus_mm.py (hunk 4 above)
Comment (blancsw) [reply vllm/model_executor/models/apertus_mm.py:450]: Great, I was searching for a better and clearer method to resolve paths for the audio and vision weights. Thanks for the tips!

File: vllm/model_executor/models/apertus_mm.py (hunk 5 above)
Comment (blancsw) [reply vllm/model_executor/models/apertus_mm.py]: perfect !

**Comment by blancsw** [comment]
Thank you for the feedback, @Isotr0py. This allows me to build my expertise on the vLLM contribution side!

**Comment by blancsw** [comment]
I close this PR, the new one can be found here
https://github.com/vllm-project/vllm/pull/50496
